eventrouter/internal/queue_posix.c

```diff
@@ -0,0 +1,237 @@
+{
+    ErEvent_t* result = a_queue->m_data[a_queue->m_idx];
+    a_queue->m_idx    = (a_queue->m_idx + 1) % a_queue->m_capacity;
+    a_queue->m_size -= 1;
+    return result;
+}
+
+static void write(Queue_t* a_queue, ErEvent_t* a_event)
+{
+    int write_idx = (a_queue->m_idx + a_queue->m_size) % a_queue->m_capacity;
+    a_queue->m_data[write_idx] = a_event;
+    a_queue->m_size += 1;
+}
+
+/// Returns a `struct timespec` corresponding to the time `a_ms` in the future.
+static struct timespec future(int64_t a_ms)
+{
+    struct timeval tv;
+    gettimeofday(&tv, NULL);
+    const int64_t nanos = (tv.tv_sec * 1e9) + (a_ms * 1e6) + (tv.tv_usec * 1e3);
+
+    struct timespec ts;
+    ts.tv_sec  = nanos / 1000000000;
+    ts.tv_nsec = nanos % 1000000000;
+    return ts;
+}
+
+//==============================================================================
+// Public Functions
+//==============================================================================
+
+ErQueue_t ErQueueNew(size_t a_capacity)
+{
+    Queue_t* const result =
+        malloc(sizeof(ErQueue_t) + (sizeof(result->m_data[0]) * a_capacity));
+
+    pthread_mutex_init(&result->m_mutex, NULL);
+    pthread_cond_init(&result->m_cond, NULL);
+    result->m_idx      = 0;
+    result->m_size     = 0;
+    result->m_capacity = a_capacity;
+
+    return result;
+}
+
+void ErQueueFree(ErQueue_t a_queue)
+{
+    assert(a_queue != NULL);
+    Queue_t* q = a_queue;
+    pthread_cond_destroy(&q->m_cond);
+    pthread_mutex_destroy(&q->m_mutex);
+    free(q);
+}
+
+ErEvent_t* ErQueuePopFront(ErQueue_t a_queue)
+{
+    assert(a_queue != NULL);
+
+    Queue_t* q        = a_queue;
+    ErEvent_t* result = 0;
+
+    // Block until there is data in the queue.
+    while (1)
+    {
+        // Check the condition and while holding the mutex to avoid a race.
+        pthread_mutex_lock(&q->m_mutex);
+        if (q->m_size == 0)
+        {
+            pthread_cond_wait(&q->m_cond, &q->m_mutex);
+        }
+        // At this point one of two things is true:
+        //
+        //   1. m_size was greater than zero.
+        //   2. m_size was initially zero and then the condition was signaled.
+        //
+        // Since we don't know which happened we must check the condition again;
+        // `pthread_cond_signal()` can wake more than one thread on some
+        // systems, and another thread could have read the data first.
+        //
+        // If there is data in the queue consume it, release the lock and break
+        // out of the loop. If there isn't any data (because another thread read
+        // it) then go back to waiting on the condition variable.
+        if (q->m_size > 0)
+        {
+            result = read(a_queue);
+            pthread_mutex_unlock(&q->m_mutex);
+            pthread_cond_signal(&q->m_cond);  // Notify blocked writers.
+            break;
+        }
+        else
+        {
+            pthread_mutex_unlock(&q->m_mutex);
+        }
+    }
```

Comment:
This is my first time using condition variables to this degree; please let me know if I've used them correctly.

Reply:
Is there a specific reason to `unlock` before `signal`?

https://pubs.opengroup.org/onlinepubs/007904975/functions/pthread_cond_broadcast.html
If predictable scheduling behavior is required, then that mutex shall be locked by the thread calling pthread_cond_broadcast() or pthread_cond_signal().